PXC-2331 : run mysql_upgrade as part of SST

kennt-percona opened the pull request:
Now running mysql_upgrade as part of the SST.  This makes it easier
to upgrade (since the donor would be a lower version and the joiner
would be a higher version).

kennt-percona commented:
https://jenkins.percona.com/job/pxc80.test.mtr/

kbauskar commented on build-ps/build-binary.sh:
```diff
@@ -328,23 +328,44 @@ fi
         export CC=${GALERA_CC:-gcc}
         export CXX=${GALERA_CXX:-g++}
 
+        # Look for boost_program_options static library
```
Please add a comment why this is needed.

kennt-percona replied:
added a comment

kbauskar commented on build-ps/build-binary.sh:
```diff
@@ -443,6 +464,75 @@ fi
+    pxb8_version=$(echo $pxb_dir | grep -oe "[1-9]\.[0-9][0-9]*\.[0-9][0-9]*")
+
+    mkdir -p "$TARGETDIR/usr/local/$PRODUCT_FULL_NAME/bin/pxc_extra/pxb-2.4"
+    (cp -v -r $TARGETDIR/pxc_extra/pxb-2.4/bin/  $TARGETDIR/usr/local/$PRODUCT_FULL_NAME/bin/pxc_extra/pxb-2.4) || true
```
Leave a comment that we explicitly copy only binary and lib. (test cases and other files are not copied).

kennt-percona replied:
added a comment

kbauskar commented on mysql-test/suite/galera/r/galera_as_master_gtid.result:
```diff
@@ -23,18 +23,7 @@ uuids_match
-mysqld-bin.000001	798	Table_map	2	846	table_id: # (test.t1)
-mysqld-bin.000001	846	Write_rows	2	886	table_id: # flags: STMT_END_F
-mysqld-bin.000001	886	Xid	2	917	COMMIT /* xid=# */
+mysqld-bin.000001	155	Stop	2	178	
```
Since server is being restarted mysqld-bin.000001 will be closed and instead, I presume mysqld-bin.000002 will have needed set of events.

kbauskar commented on mysql-test/suite/galera/r/galera_as_slave_gtid.result:
```diff
@@ -39,13 +39,7 @@ gtid_executed_equal
-mysqld-bin.000001	531	Table_map	1	579	table_id: # (test.t1)
-mysqld-bin.000001	579	Write_rows	1	619	table_id: # flags: STMT_END_F
-mysqld-bin.000001	619	Xid	1	650	COMMIT /* xid=2 */
+mysqld-bin.000001	155	Stop	3	178	
```
Same as above.

kbauskar commented on scripts/wsrep_sst_xtrabackup-v2.sh:
```diff
@@ -459,6 +462,7 @@ get_transfer()
             if [[ -z "$SOCAT_VERSION" ]]; then
                 wsrep_log_error "******************* FATAL ERROR ********************** "
                 wsrep_log_error "* Cannot determine the socat version. "
+                wsrep_log_error "* Line $LINENO"
```
wondering what is the significance of printing LINENO.

kennt-percona replied:
Makes it easier to locate the error within the code when debugging

kbauskar replied:
ok. We can keep it but otherwise error message that is printed anyway helps to trace respective line.

kbauskar commented on scripts/wsrep_sst_xtrabackup-v2.sh:
```diff
@@ -1346,32 +1496,30 @@ fi
+#
+XB_2x_VERSION=$($XTRABACKUP_24_PATH/bin/$XTRABACKUP_BIN --version 2>&1 | grep -oe ' [0-9]\.[0-9]\.[0-9]*' | head -n1)
+XB_2x_VERSION=${XB_2x_VERSION# }
+XB_2x_REQUIRED_VERSION="2.4.13"
```
Why is 2.4.13 ?

kennt-percona replied:
Just the latest version released.

kbauskar replied:
We tagged to 2.4.12 since it has some important fixes that we banked on. like keyring and all. There is no such significance of 2.4.13 but since this is the first release for 8.0 we can continue to tag it to 2.4.13 being the latest one and we plan to ship release as part of PXC.

kbauskar commented on scripts/wsrep_sst_xtrabackup-v2.sh:
```diff
@@ -1346,32 +1496,30 @@ fi
-if ! check_for_version $XB_VERSION $XB_REQUIRED_VERSION; then
+XB_8x_VERSION=$($XTRABACKUP_80_PATH/bin/$XTRABACKUP_BIN --version 2>&1 | grep -oe ' [0-9]\.[0-9]\.[0-9]*' | head -n1)
+XB_8x_VERSION=${XB_8x_VERSION# }
+XB_8x_REQUIRED_VERSION="8.0.4"
```
Remember to update this to 8.0.5 once it is released.

kbauskar reviewed (approved):
do check to fix the mtr test-cases.

kennt-percona committed:
PXC-2331 : run mysql_upgrade as part of SST

Now running mysql_upgrade as part of the SST.  This makes it easier
to upgrade (since the donor would be a lower version and the joiner
would be a higher version).
